server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java

```diff
@@ -651,14 +651,22 @@ private static void parseObjectDynamic(DocumentParserContext context, String cur
         if (context.dynamic() == ObjectMapper.Dynamic.FALSE) {
             failIfMatchesRoutingPath(context, currentFieldName);
             if (context.canAddIgnoredField()) {
-                // read everything up to end object and store it
+                // Don't expand dots while storing the object.
+                // It is possible that it is not fully normalized (e.g. has dotted names like obj.field)
+                // which is a valid document but it is not a valid json when parsed with expanded dots
+                // due to duplicate keys.
+                // F.e. { "a.b": "b", "a.c": "c" } will be transformed into { "a": { "b": "b" }, "a": { "c": "c" } }
+                context.path().setWithinLeafObject(true);
+
```

Comment:
I am wondering if calling `setWithinLeafObject` should actually be done inside `addIgnoredField`. Calling it is required to make ignore logic work not really because of `DocumentParser`...so I think it is a responsibility of the "ignore logic".

Reply:
Yes, this is a very good point. I was not sure what is the best structure so i opened a draft with this.